添加英文版的 ONNX 文章 (#435)
Oneflow-Inc/oneflow-documentation: en/docs/cookies/oneflow2onnnx.md

This document introduces the usage of OneFlow interacting with ONNX, including an introduction to ONNX, how to export OneFlow models to ONNX models, and how to use ONNX models for inference.

## Introduction to ONNX

[ONNX](https://onnx.ai/index.html), known as Open Neural Network Exchange, is an open file format standard designed for machine learning algorithms to store trained algorithmic models. Many major deep learning frameworks (e.g., OneFlow, PyTorch, TensorFlow, MXNet) support exporting models to ONNX models, which allows different deep learning frameworks to store and interact with model data in a uniform format. In addition, ONNX has a corresponding Runtime - [ONNX Runtime](https://onnxruntime.ai/) - that facilitates model deployment and reasoning on multiple platforms (Linux, Windows, Mac OS, Android, iOS, etc.) and multiple hardware (CPU, GPU, etc.). 

### ONNX-Related Libraries

There are several ONNX-related libraries, and the features of several common libraries are described below. The onnxruntime-gpu involved in this tutorial can be installed via `pip install onnxruntime-gpu`.

1. [onnx](https://github.com/onnx/onnx): ONNX model format standard

> 2. [onnxruntime & onnxruntime-gpu](https://github.com/microsoft/onnxruntime): ONNX runtime that is used to load the ONNX model for inference. ?onnxruntime and onnxruntime-gpu support CPU inference and GPU inference respectively.

Author: 这里的onnxruntime首字母可以大写吗

Reviewer: 这里几个是作为包名的，包名确实是小写，所以目前这样就可以了。

Author: 好的

> 2. [onnxruntime & onnxruntime-gpu](https://github.com/microsoft/onnxruntime): ONNX runtime that is used to load the ONNX model for inference. ?onnxruntime and onnxruntime-gpu support CPU inference and GPU inference respectively.

Author: ```suggestion
2. [onnxruntime & onnxruntime-gpu](https://github.com/microsoft/onnxruntime): ONNX runtime that is used to load the ONNX model for inference. onnxruntime and onnxruntime-gpu support CPU inference and GPU inference respectively.
```

   
4. [onnxoptimizer](https://github.com/onnx/optimizer): it is used to optimize ONNX model through graph transformations

## Export OneFlow Models to ONNX Models

[oneflow-onnx](https://github.com/Oneflow-Inc/oneflow_convert) is a model conversion tool provided by OneFlow team to support exporting OneFlow static diagram models to ONNX models. At present oneflow-onnx supports more than 80 kinds of Oneflow OP exported as ONNX OP. For detalis, refer to [list of OP supported by OneFlow2ONNX](https://github.com/Oneflow-Inc/oneflow_convert/blob/main/docs/oneflow2onnx/op_list.md)。

### Install oneflow-onnx

oneflow-onnx is independent of OneFlow and needs to be installed separately via pip:
```bash
pip install oneflow-onnx
```

### How to Use oneflow-onnx

To export OneFlow static graph model as ONNX model, just call ` export_ onnx_ Model ` function.

```python
from oneflow_onnx.oneflow2onnx.util import export_onnx_model

export_onnx_model(graph,
                  external_data=False, 
                  opset=None, 
                  flow_weight_dir=None, 
                  onnx_model_path="/tmp", 
                  dynamic_batch_size=False)

> ```

Author: ```suggestion
```

```

> ```

Author: ```suggestion
```

```

> The meaning of each parameter is as follows:

Author: ```suggestion

The meaning of each parameter is as follows:
```

> The meaning of each parameter is as follows:

Author: 已修改

> 1. graph: the graph need to be converted ( [Graph](../basics/08_nn_graph.md) object )

Author: 中文版留了空格 所以我将英文版也留了 我去找晓祯确认下

2. external_data: Whether to save the weights as external data of the ONNX model. When it is `True`, it is usually to avoid the 2GB file size limit of protobuf.

3. opset: Specify the version of the conversion model ( int, default is 10 )

4. flow_weight_dir: Path to save OneFlow model weights

5. onnx_model_path: Save path for exported ONNX models

> 6. dynamic_batch_size: Whether the exported ONNX model supports dynamic batch, default is False

Author: 列举部分冒号后首字母是否需要大写，句子末尾是否需要加句号

> In addition, oneflow-onnx provides a function called `convert_to_onnx_and_check` to convert and check the converted ONNX model. The check refers to feeding the same input to the original OneFlow model and the converted ONNX model respectively, and then comparing the difference between each value in the two outputs to see if it is within a reasonable margin of error.

Reviewer: ```suggestion
In addition, oneflow-onnx provides a function called `convert_to_onnx_and_check` to convert and meanwhile check the converted ONNX model. The check process will pass the same input to the original OneFlow model and the converted ONNX model respectively, and then compare the difference between each value in the two outputs to see if they are same within a relative range.
```

```python
from oneflow_onnx.oneflow2onnx.util import convert_to_onnx_and_check

convert_to_onnx_and_check(...)
```

> The parameters of the `convert_to_onnx_and_check` are a superset of the parameters of the `export_onnx_model`, and you can be pass in `print_outlier=True` additionally to output any abnormal values found during the check process that exceed the reasonable error range. 

Author: ```suggestion
The parameters of the `convert_to_onnx_and_check` are almost the same as those of `export_onnx_model`, besides  you can pass `print_outlier` parameter additionally. When `print_outlier=True`, it will output any abnormal values found during the check process that exceed the reasonable error range. 
```

> The parameters of the `convert_to_onnx_and_check` are a superset of the parameters of the `export_onnx_model`, and you can be pass in `print_outlier=True` additionally to output any abnormal values found during the check process that exceed the reasonable error range. 

Reviewer: 这里不用直译。可以这样表述：

convert_to_onnx_and_check 的参数和 export_onnx_model 几乎是一样的，besides 可以多传递一个 print_outlier 参数。当 `print_outlier=True` 时，……

你来整理成英文吧。

### Considerations when Exporting Models

> - Before exporting the model, you need to set the model to eval mode because operations such as Dropout and Batch Normalization have different behaviors under the training and evaluation mode.

Author: ```suggestion
- Before exporting the model, it need be set to eval mode because operations such as Dropout and Batch Normalization have different behaviors under the training and evaluation mode.
```

> - Before exporting the model, you need to set the model to eval mode because operations such as Dropout and Batch Normalization have different behaviors under the training and evaluation mode.

Reviewer: 这里的 you 其实是不必要出现的，可以改成

model need be set to eval model before exproted ..., because ....

具体的语法细节麻烦修改时自己确认好吧

> - When building a static graph model, you need to specify an input. The value of this input can be random, but make sure it is the correct data type and shape.

Reviewer: ```suggestion
- When building a static graph model, you need to specify an input. The value of the input can be random, but make sure the data type and shape is correct.
```

原来的写法比较中式英语吧？

> - The ONNX model accepts a fixed shape of input, and the varied size of the batch dimension, so by setting the `dynamic_batch_size` parameter to be `True` you can make the exported ONNX model support dynamic batch size.

Reviewer: ```suggestion
- The ONNX model accepts a fixed shape of input, and a varied size of the batch dimension, so by setting the `dynamic_batch_size` parameter to be `True` can make the exported ONNX model support dynamic batch size.
```

或者用 and the size of batch dimension can varied

you 也可以去掉

英文中手册突然出现 you 会显得很突兀，感觉在被人命令一样。这个可以多阅读其它的用户手册来形成语感

- Oneflow-onnx must use a static graph model (Graph mode) as an parameter to export function. For dynamic graph models (Eager mode), the dynamic graph model needs to be constructed as a static graph model. Refer to the example below.

## Examples of Usage

In this section, the process of exporting a OneFlow model to an ONNX model and performing inference is introduced, using the common ResNet-34 model as an example.

The following code uses [FlowVision](https://github.com/Oneflow-Inc/vision), a library built on OneFlow for computer vision tasks, which contains many models, data enhancement methods, data transformation operations, datasets, and so on. Here we directly use the ResNet-34 model provided by the FlowVision library and use its weight trained on the ImageNet dataset. 

### Export as ONNX Model

> Import related dependencies:

Author: ```suggestion
Import related dependencies:

```

> ```python

Reviewer: 上面加个空行。

Author: 已修改

import oneflow as flow
from oneflow import nn
from flowvision.models import resnet34
from oneflow_onnx.oneflow2onnx.util import convert_to_onnx_and_check
```

> To build a static graph model using a dynamic graph model. For details, refer to: [Static Graph Module nn.Graph](../basics/08_nn_graph.md)

Reviewer: ```suggestion
To build a static graph model using a dynamic graph model. For details, refer to: [Static Graph Interface: nn.Graph](../basics/08_nn_graph.md)
```

我看了下，那篇文章的标题不是 Graph Module

```python
class ResNet34Graph(nn.Graph):
    def __init__(self, eager_model):
        super().__init__()
        self.model = eager_model

    def build(self, x):
        return self.model(x)
```

> Export OneFlow static graph models to ONNX models:
> ```python

Reviewer: ```suggestion
Export OneFlow static graph models to ONNX models:

```python
```

# Model parameter storage directory
MODEL_PARAMS = 'checkpoints/resnet34'

params = flow.load(MODEL_PARAMS)
model = resnet34()
model.load_state_dict(params)

# Set the model to eval mode
model.eval()

resnet34_graph = ResNet34Graph(model)
# Build the static graph model
resnet34_graph._compile(flow.randn(1, 3, 224, 224))

# Export as ONNX model and check
convert_to_onnx_and_check(resnet34_graph, 
                          flow_weight_dir=MODEL_PARAMS, 
                          onnx_model_path="./", 
                          print_outlier=True,
                          dynamic_batch_size=True)
```

> After running, you can find a file named `model.onnx` in the current directory, which is the exported ONNX model.

Author: ```suggestion
After running,  a file named `model.onnx`is  in the current directory, which is the exported ONNX model.
```

> ```
> After running, you can find a file named `model.onnx` in the current directory, which is the exported ONNX model.

Reviewer: ```suggestion
```

After running, you can find a file named `model.onnx` in the current directory, which is the exported ONNX model.
```

Reviewer: 改变句式，去掉 you

### Inference with ONNX models

> Before performing inference, ensure that the ONNX Runtime is installed, that is onnxruntime or onnxruntime-gpu. In the experimental environment of this tutorial, onnxruntime-gpu is installed to invoke the GPU for calculation(or computation??不确定), but if there is no GPU on the machine, you can specify the CPU for calculation. See below for details.

Author: 这里的使用“GPU计算”使用calculation还是computation

Reviewer: 用 computation 吧，没什么特别的，就是之前用得多，同意下。而且 compute 更加 general 一点

Author: 已修改

Reviewer: > Before performing inference, ensure that the ONNX Runtime is installed, that is onnxruntime or onnxruntime-gpu.

这类句子感觉非常的古怪。要能够自己分辨出它的古怪……

ONNX Runtime(onnxruntime or onnxruntime-gpu) is required for inference.

Reviewer: > In the experimental environment of this tutorial,

In the environment of this tutorial, 即可

> Before performing inference, ensure that the ONNX Runtime is installed, that is onnxruntime or onnxruntime-gpu. In the experimental environment of this tutorial, onnxruntime-gpu is installed to invoke the GPU for calculation(or computation??不确定), but if there is no GPU on the machine, you can specify the CPU for calculation. See below for details.

Author: ```suggestion
Before performing inference, ensure that the ONNX Runtime is installed, that is onnxruntime or onnxruntime-gpu. In the experimental environment of this tutorial, onnxruntime-gpu is installed to invoke the GPU for computation, but if there is no GPU on the machine, you can specify the CPU for calculation. See below for details.
```

import cv2
from onnxruntime import InferenceSession
```

Define a function to pre-process the image to a format and size accepted by the ONNX model:
```python
def preprocess_image(img, input_hw = (224, 224)):
    h, w, _ = img.shape
    
    # Use the longer side of the image to determine the scaling factor
    is_wider = True if h <= w else False
    scale = input_hw[1] / w if is_wider else input_hw[0] / h

    # Scale the image equally
    processed_img = cv2.resize(img, (0, 0), fx=scale, fy=scale, interpolation=cv2.INTER_LINEAR)
    # Normalization
    processed_img = np.array(processed_img, dtype=np.float32) / 255

    # Fill images to ONNX model and preset sizes
    temp_img = np.zeros((input_hw[0], input_hw[1], 3), dtype=np.float32)
    temp_img[:processed_img.shape[0], :processed_img.shape[1], :] = processed_img
    processed_img = temp_img
    
    # Adjust the order of axes and add batch axes at the front 
    processed_img = np.expand_dims(processed_img.transpose(2, 0, 1), axis=0)

    return processed_img
```

The next step is to use the ONNX model for inference, which consists of creating an InferenceSession object and calling `run` to perform inference.

> In onnxruntime(-gpu) 1.9 and above, the `providers` parameter needs to be explicitly specified when creating an InferenceSession object to select the hardware to use. For onnxruntime-gpu, the values that can be specified include `TensorrtExecutionProvider`, `CUDAExecutionProvider`, and `CPUExecutionProvider`. If there is no GPU on the running machine, you can specify the `providers` parameter as `['CPUExecutionProvider']` to use the CPU for computation？.

Author: 这里的使用“CPU计算”使用calculation还是computation

Reviewer: 同上，用 computation 吧

Author: 已修改

> In onnxruntime(-gpu) 1.9 and above, the `providers` parameter needs to be explicitly specified when creating an InferenceSession object to select the hardware to use. For onnxruntime-gpu, the values that can be specified include `TensorrtExecutionProvider`, `CUDAExecutionProvider`, and `CPUExecutionProvider`. If there is no GPU on the running machine, you can specify the `providers` parameter as `['CPUExecutionProvider']` to use the CPU for computation？.

Author: ```suggestion
In onnxruntime(-gpu) 1.9 and above, the `providers` parameter needs to be explicitly specified when creating an InferenceSession object to select the hardware to use. For onnxruntime-gpu, the values that can be specified include `TensorrtExecutionProvider`, `CUDAExecutionProvider`, and `CPUExecutionProvider`. If there is no GPU on the running machine, you can specify the `providers` parameter as `['CPUExecutionProvider']` to use the CPU for computation.
```

> The type of input data of an ONNX model is a dict. Its keys are `input names` when exporting the ONNX model, and the values are the actual input data of NumPy array type. You can get "input names" through the `get_inputs` method of the InferenceSession object, which returns a list of objects of `onnxruntime.NodeArg` type. For NodeArg object, you can use its `name` property to get a name of str type. In this tutorial, the input is only the image data, so you can get the "input names" corresponding to the input by calling `.get_inputs()[0].name` on the InferenceSession object. The value is `_ResNet34Graph_0-input_0/out`, which is used as the key to construct the dict of the ONNX model input. Of course, it can also be obtained dynamically at runtime without specifying it in advance.

Reviewer: 此处往下还没开始看。

# Read the image file and preprocess it with the `preprocess_image` function
img = cv2.imread('cat.jpg', cv2.IMREAD_COLOR)
img = preprocess_image(img)

# Create an InferenceSession object
ort_sess = InferenceSession('model.onnx', providers=['TensorrtExecutionProvider',
                                                     'CUDAExecutionProvider',
                                                     'CPUExecutionProvider'])
# Call the `run` method of the InferenceSession object to perform inference
results = ort_sess.run(None, {"_ResNet34Graph_0-input_0/out": img})

# Output inference results
print(CLASS_NAMES[np.argmax(results[0])])
```

The output of the `run` method of the InferenceSession object is a list of NumPy arrays, and each NumPy array corresponds to a set of outputs. Since there is only one set of inputs, the element with index 0 is the output, and the shape of it is `(1, 1000)`, which corresponds to the probability of 1000 categories (if n images are input as a batch, the shape of them will be `(n, 1000)`). After obtaining the index corresponding to the category with the highest probability via `np.argmax`, the index is mapped to the category name.

> Run the above code:

Reviewer: ```suggestion
Run the code and get the result:
```

                  onnx_model_path="/tmp", 
                  dynamic_batch_size=False)
```

The meaning of each parameter is as follows:

1. graph: the graph need to be converted ( [Graph](../basics/08_nn_graph.md) object )

2. external_data: whether to save the weights as external data of the ONNX model. When it is `True`, it is usually to avoid the 2GB file size limit of protobuf.

3. opset: specify the version of the conversion model ( int, default is 10 )

4. flow_weight_dir: path to save OneFlow model weights

5. onnx_model_path: save path for exported ONNX models

6. dynamic_batch_size: whether the exported ONNX model supports dynamic batch, default is False

In addition, oneflow-onnx provides a function called `convert_to_onnx_and_check` to convert and meanwhile check the converted ONNX model. The check process will pass the same input to the original OneFlow model and the converted ONNX model respectively, and then compare the difference between each value in the two outputs to see if they are same within a relative range.

```python
from oneflow_onnx.oneflow2onnx.util import convert_to_onnx_and_check

convert_to_onnx_and_check(...)
```

> The parameters of the `convert_to_onnx_and_check` are almost the same as those of `export_onnx_model`, besides  you can pass `print_outlier` parameter additionally. When `print_outlier=True`, it will output any abnormal values found during the check process that exceed the reasonable error range. 

Author: ```suggestion

The parameters of the `convert_to_onnx_and_check` are almost the same as those of `export_onnx_model`, besides  you can pass `print_outlier` parameter additionally. When `print_outlier=True`, it will output any abnormal values found during the check process that exceed the reasonable error range. 
```

```python
import oneflow as flow
from oneflow import nn
from flowvision.models import resnet34
from oneflow_onnx.oneflow2onnx.util import convert_to_onnx_and_check
```

To build a static graph model using a dynamic graph model. For details, refer to: [Static Graph Interface: nn.Graph](../basics/08_nn_graph.md)

```python
class ResNet34Graph(nn.Graph):
    def __init__(self, eager_model):
        super().__init__()
        self.model = eager_model

    def build(self, x):
        return self.model(x)
```

Export OneFlow static graph models to ONNX models:

```python
# Model parameter storage directory
MODEL_PARAMS = 'checkpoints/resnet34'

params = flow.load(MODEL_PARAMS)
model = resnet34()
model.load_state_dict(params)

# Set the model to eval mode
model.eval()

resnet34_graph = ResNet34Graph(model)
# Build the static graph model
resnet34_graph._compile(flow.randn(1, 3, 224, 224))

# Export as ONNX model and check
convert_to_onnx_and_check(resnet34_graph, 
                          flow_weight_dir=MODEL_PARAMS, 
                          onnx_model_path="./", 
                          print_outlier=True,
                          dynamic_batch_size=True)
```

> After running,  a file named `model.onnx` is  in the current directory, which is the exported ONNX model.

Author: ```suggestion

After running,  a file named `model.onnx` is  in the current directory, which is the exported ONNX model.
```

### Inference with ONNX models

Before performing inference, ensure that the ONNX Runtime is installed, that is onnxruntime or onnxruntime-gpu. In the experimental environment of this tutorial, onnxruntime-gpu is installed to invoke the GPU for computation, but if there is no GPU on the machine, you can specify the CPU for calculation. See below for details.

We use the following image as input to the model:
<div align="center">
    <img alt="Demo Image" src="./imgs/cat.jpg" width="300px">
</div>

> Import related dependencies:

Author: ```suggestion
Import related dependencies:

```

```python
import numpy as np
import cv2
from onnxruntime import InferenceSession
```

> Define a function to pre-process the image to a format and size accepted by the ONNX model:

Author: ```suggestion
Define a function to pre-process the image to a format and size accepted by the ONNX model:

```

# Read the image file and preprocess it with the `preprocess_image` function
img = cv2.imread('cat.jpg', cv2.IMREAD_COLOR)
img = preprocess_image(img)

# Create an InferenceSession object
ort_sess = InferenceSession('model.onnx', providers=['TensorrtExecutionProvider',
                                                     'CUDAExecutionProvider',
                                                     'CPUExecutionProvider'])
# Call the `run` method of the InferenceSession object to perform inference
results = ort_sess.run(None, {"_ResNet34Graph_0-input_0/out": img})

# Output inference results
print(CLASS_NAMES[np.argmax(results[0])])
```

The output of the `run` method of the InferenceSession object is a list of NumPy arrays, and each NumPy array corresponds to a set of outputs. Since there is only one set of inputs, the element with index 0 is the output, and the shape of it is `(1, 1000)`, which corresponds to the probability of 1000 categories (if n images are input as a batch, the shape of them will be `(n, 1000)`). After obtaining the index corresponding to the category with the highest probability via `np.argmax`, the index is mapped to the category name.

> Run the code and get the result:

Author: ```suggestion
Run the code and get the result:

```

                  onnx_model_path="/tmp", 
                  dynamic_batch_size=False)
```

The meaning of each parameter is as follows:

1. graph: the graph need to be converted ([Graph](../basics/08_nn_graph.md) object)

2. external_data: whether to save the weights as external data of the ONNX model. When it is `True`, it is usually to avoid the 2GB file size limit of protobuf.

3. opset: specify the version of the conversion model (int, default is 10)

4. flow_weight_dir: path to save OneFlow model weights

5. onnx_model_path: save path for exported ONNX models

6. dynamic_batch_size: whether the exported ONNX model supports dynamic batch, default is False

In addition, oneflow-onnx provides a function called `convert_to_onnx_and_check` to convert and meanwhile check the converted ONNX model. The check process will pass the same input to the original OneFlow model and the converted ONNX model respectively, and then compare the difference between each value in the two outputs to see if they are same within a relative range.

```python
from oneflow_onnx.oneflow2onnx.util import convert_to_onnx_and_check

convert_to_onnx_and_check(...)
```

> The parameters of the `convert_to_onnx_and_check` are almost the same as those of `export_onnx_model`, besides  you can pass `print_outlier` parameter additionally. When `print_outlier=True`, it will output any abnormal values found during the check process that exceed the reasonable error range. 

Reviewer: ```suggestion
The parameters of the `convert_to_onnx_and_check` are almost the same as those of `export_onnx_model`, besides you can pass `print_outlier` parameter additionally. When `print_outlier=True`, it will output any abnormal values found during the check process that exceed the reasonable error range. 
```
